File: apps/desktop/src/mappers/spending.ts
import type { SpendingResponse, SpendingTransactionResponse } from '@nova/api-contracts'
import type { SpendingEntry, SpendingViewModel } from '@/view-models'

export function mapSpendingTransaction(row: SpendingTransactionResponse): SpendingEntry {
  return {
    id: row.id,
    type: row.type,
    amount: row.amount,
    note: row.note ?? undefined,
    createdAt: row.created_at,
  }
}

export function mapSpending(response: SpendingResponse): SpendingViewModel {
  return {
    earnedMonth: response.earned_month,
    spentMonth: response.spent_month,
    transactions: response.transactions.map(mapSpendingTransaction),
  }
}

File: apps/backend/services/api/schemas/spending.py
from typing import Literal, Optional

from pydantic import BaseModel, Field

from .common import MutationMeta


class SpendingTransactionResponse(BaseModel):
    id: int
    type: str
    amount: float
    note: Optional[str] = None
    created_at: str


class SpendingSummaryResponse(BaseModel):
    period: str
    earned: float
    spent: float
    summary: str


class SpendingResponse(BaseModel):
    earned_month: float
    spent_month: float
    transactions: list[SpendingTransactionResponse]


class LogSpendingRequest(BaseModel):
    type: Literal["earned", "spent"]
    amount: float = Field(..., gt=0)
    note: Optional[str] = None


class SpendingMutationResponse(BaseModel):
    item: SpendingTransactionResponse
    meta: MutationMeta
